Distributed training not working for PEFT/AdaLoRA whisper

rrva opened the pull request:
These are some changes that seem to be needed for multi-gpu training.

However, training still fails:

```
Traceback (most recent call last):
  File "/root/peft/examples/int8_training/peft_adalora_whisper_large_training.py", line 794, in <module>
    main()                
  File "/root/peft/examples/int8_training/peft_adalora_whisper_large_training.py", line 708, in main
    get_base_model(model).update_and_allocate(global_step)
  File "/root/env/lib/python3.10/site-packages/peft/tuners/adalora.py", line 388, in update_and_allocate
    self.rankallocator.mask_using_rank_pattern(self.model, lora_config.rank_pattern)      
  File "/root/env/lib/python3.10/site-packages/peft/tuners/adalora.py", line 815, in mask_using_rank_pattern
    if self.adapter_name not in next(iter(rank_pattern.keys())):
AttributeError: 'NoneType' object has no attribute 'keys'
```

It seems that `lora_config.rank_pattern` becomes `None` in this code:

```python
def update_and_allocate(self, global_step):
    lora_config = self.peft_config[self.trainable_adapter_name]
    # Update the importance score and allocate the budget
    if global_step < lora_config.total_step - lora_config.tfinal:
        _, rank_pattern = self.rankallocator.update_and_allocate(self.model, global_step)
        if rank_pattern:
            lora_config.rank_pattern = rank_pattern
    # Finalize the budget allocation
    elif global_step == lora_config.total_step - lora_config.tfinal:
        _, rank_pattern = self.rankallocator.update_and_allocate(self.model, global_step, force_mask=True)
        # for some reason, this freezes the trainable parameters and nothing gets updates
        # self.resize_modules_by_rank_pattern(rank_pattern, self.trainable_adapter_name)
        lora_config.rank_pattern = rank_pattern
        self.rankallocator.reset_ipt()
    # Currently using inefficient way to mask the unimportant weights using the rank pattern
    #  due to problem mentioned above
    elif global_step > lora_config.total_step - lora_config.tfinal:
        self.rankallocator.mask_using_rank_pattern(self.model, lora_config.rank_pattern)
    # Pass the function and do forward propagation
    else:
        return None
``` 

Which I have no idea why.

rrva commented:
Some of this code is adapted from https://github.com/huggingface/peft/issues/242

@younesbelkada @pacman100 

Any idea why lora_config.rank_pattern becomes None?

rrva commented on examples/int8_training/peft_adalora_whisper_large_training.py:
```diff
@@ -326,11 +331,18 @@ def save_model_hook(models, weights, output_dir):
         weights.pop()
 
 
+def get_base_model(model):
```
This is because model becomes a DistributedDataParallel when training in distributed mode. Not sure if this is the correct/intended way to do it, but at least it abstracts it a bit

rrva commented on examples/int8_training/peft_adalora_whisper_large_training.py:
```diff
@@ -538,7 +550,16 @@ def main():
 
     # model
-    model = WhisperForConditionalGeneration.from_pretrained(args.model_name_or_path, load_in_8bit=True)
+    if (args.multi_gpu):
```
This code is taken from https://github.com/huggingface/peft/issues/242#issuecomment-1491447956

rrva commented on examples/int8_training/peft_adalora_whisper_large_training.py:
```diff
@@ -684,7 +705,7 @@ def make_inputs_require_grad(module, input, output):
                 # Hence being called before optimizer.zero_grad().
                 if args.use_peft and args.use_adalora:
-                    model.update_and_allocate(global_step)
+                    get_base_model(model).update_and_allocate(global_step)
```
beyond this step I still get problems further down with lora rank_pattern

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_855). All of your documentation changes will be reflected on that endpoint.

younesbelkada reviewed (commented):
Hi @rrva thanks a lot for your PR, I will let @pacman100 decide about this PR as he wrote the FT script for Whisper, can you also run the styling checks?
```bash
make style && make quality
```

rrva commented:
> Hi @rrva thanks a lot for your PR, I will let @pacman100 decide about this PR as he wrote the FT script for Whisper, can you also run the styling checks?
> 
> ```shell
> make style && make quality
> ```

fixed the style issues and rebased

rrva commented:
@pacman100 seen this? I think it just needs minor adjustments

IliyanGochev commented:
@rrva 

I've started using this script and I think I hit the same problems as you. 
If you look at https://github.com/huggingface/transformers/issues/18974 you can have an idea how to fix the problems.

You'd need to do things like:
`
PeftModel.from_pretrained(model.module.base_model.model, input_dir)
`

and replace `model.generate(...)` with  ` accelerator.unwrap_model(model).generate(...) `

There are several places you'd need to make the changes. 
Then run the script with `accelerate launch peft_adalora_whisper_large_training.py --args.... `
It will default to all GPUs on the current machine. Right now I'm applying PEFT for Whisper large for Bulgarian fine-tuning. Be sure to specify an output dir, it won't be created automatically as the argument is None.

I'll try to prepare a PR for the script in the next few days, after a successful run. I'm using the latest PyPI versions of PEFT, Accelerate, Transformers.

KarwanKhalid commented:
@rrva  I am still encountering the same error. Did you manage to fix it?
